File: ios/LynxShellKit/Routing/LynxRouteParser.swift
import Foundation

/**
 * 兼容 lynxshell、Sparkling hybrid、Explorer 本地地址、HTTPS 与 App Bundle 路由。
 *
 * 为方便从已有 Sparkling 页面迁移，解析器同时接受 snake_case 与 camelCase 参数。
 */
enum LynxRouteParser {
    private static let explorerLocalPrefix = "file://lynx?local://"

    static func request(from route: String, optionsJSON: String = "{}") throws -> LynxPageRequest {
        let options = try decodeObject(optionsJSON, field: "options")
        let base: LynxPageRequest

        if route.contains("://") {
            base = try requestFromURL(route)
        } else {
            base = defaultRequest(bundleURL: normalizeBundle(route))
        }
        return try applying(options: options, to: base).validated()
    }

    static func request(
        bundleURL: String,
        title: String,
        initialDataJSON: String,
        globalPropsJSON: String,
        fullscreen: Bool,
        allowHTTPInDebug: Bool
    ) throws -> LynxPageRequest {
        try LynxPageRequest(
            bundleURL: normalizeBundle(bundleURL),
            routeKey: normalizeBundle(bundleURL),
            title: title.isEmpty ? "Lynx" : title,
            initialData: decodeObject(initialDataJSON, field: "initData"),
            globalProps: decodeObject(globalPropsJSON, field: "globalProps"),
            fullscreen: fullscreen,
            showNavigationBar: !fullscreen,
            hideStatusBar: false,
            backGestureEnabled: true,
            allowHTTPInDebug: allowHTTPInDebug,
            orientation: .system,
            backgroundColor: "#FFFFFF",
            widthInPhysicalPixels: nil,
            heightInPhysicalPixels: nil
        ).validated()
    }

    private static func requestFromURL(_ route: String) throws -> LynxPageRequest {
        // Explorer 的 file://lynx?local:// 不是标准 URL 结构，必须先按原字符串拆解。
        if route.lowercased().hasPrefix(explorerLocalPrefix) {
            return try requestFromExplorerLocal(route)
        }
        guard let components = URLComponents(string: route) else {
            throw LynxRouteError.unsupportedURL(route)
        }
        let scheme = components.scheme?.lowercased() ?? ""
        let host = components.host?.lowercased() ?? ""
        var query: [String: String] = [:]
        // 同名参数按最后一个值生效，避免 Dictionary(uniqueKeysWithValues:) 重复 key 崩溃。
        for item in components.queryItems ?? [] {
            query[item.name] = item.value ?? ""
        }

        if scheme == "lynxshell" || (scheme == "hybrid" && host == "lynxview_page") {
            guard let rawBundle = first(query, keys: ["url", "bundle"]), !rawBundle.isEmpty else {
                throw LynxRouteError.missingBundle
            }
            return try requestFromQuery(bundleURL: normalizeBundle(rawBundle), query: query)
        }

        if ["assets", "https", "http"].contains(scheme) {
            return try requestFromQuery(bundleURL: route, query: query)
        }
        throw LynxRouteError.unsupportedURL(route)
    }

    /**
     * `file://lynx?local://main.lynx.bundle?fullscreen=true` 的第一个问号属于协议，
     * 第二个问号才是页面参数，URLComponents 无法直接表达这层含义。
     */
    private static func requestFromExplorerLocal(_ route: String) throws -> LynxPageRequest {
        let start = route.index(route.startIndex, offsetBy: explorerLocalPrefix.count)
        let payload = String(route[start...])
        let pieces = payload.split(separator: "?", maxSplits: 1, omittingEmptySubsequences: false)
        let path = String(pieces.first ?? "").removingPercentEncoding ?? String(pieces.first ?? "")
        guard !path.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw LynxRouteError.missingBundle
        }
        let query = pieces.count > 1 ? parseEncodedQuery(String(pieces[1])) : [:]
        let bundle = explorerLocalPrefix + stripRelativePrefix(path)
        return try requestFromQuery(bundleURL: bundle, query: query)
    }

    private static func requestFromQuery(
        bundleURL: String,
        query: [String: String]
    ) throws -> LynxPageRequest {
        let fullscreen = bool(first(query, keys: ["fullscreen"]), default: true)
        let hiddenNavigation = bool(
            first(query, keys: ["hidden_nav", "hide_nav_bar", "hideNavigationBar"]),
            default: false
        )
        let explicitNavigation = first(query, keys: ["showNavigationBar", "showToolbar"])
        let showNavigationBar = explicitNavigation.map { !fullscreen && bool($0, default: true) }
            ?? (!fullscreen && !hiddenNavigation)
        let hideStatusBar = bool(
            first(query, keys: ["hide_status_bar", "hideStatusBar"]),
            default: false
        )
        let backGestureEnabled = bool(
            first(query, keys: ["backGestureEnabled", "back_gesture_enabled"]),
            default: true
        )
        let title = first(query, keys: ["title"])?.nonEmpty ?? "Lynx"
        let initialData = try decodeObject(
            first(query, keys: ["initData", "initialData", "initial_data"]) ?? "{}",
            field: "initData"
        )
        var globalProps = try decodeObject(
            first(query, keys: ["globalProps", "global_props"]) ?? "{}",
            field: "globalProps"
        )
        // Sparkling 页面通过 lynx.__globalProps.queryItems 读取当前路由来源与参数。
        globalProps["queryItems"] = query
        let allowHTTPInDebug = bool(
            first(query, keys: ["allowHTTPInDebug", "allowHttpInDebug", "allow_http_in_debug"]),
            default: false
        )
        let orientation = parseOrientation(first(query, keys: ["orientation", "screen_orientation"]))
        let backgroundColor = normalizeColor(
            first(
                query,
                keys: ["backgroundColor", "background_color", "container_bg_color"]
            ) ?? "#FFFFFF"
        )
        let widthInPhysicalPixels = first(query, keys: ["width", "width_px"])
            .flatMap { Double($0) }
            .map { CGFloat($0) }
        let heightInPhysicalPixels = first(query, keys: ["height", "height_px"])
            .flatMap { Double($0) }
            .map { CGFloat($0) }
        let requestedAppId = first(query, keys: ["lynxAppId", "appId", "lynx_app_id"])
            .flatMap { $0.trimmingCharacters(in: .whitespacesAndNewlines).nonEmpty }
        // HTTPS 始终是 Direct Remote；即使 URL query 带 appId 也不允许隐式进入 OTA。
        let otaAppId = RemoteBundlePolicy.isRemote(bundleURL) ? nil : requestedAppId
        let otaBundleName = otaAppId.flatMap { _ in
            first(query, keys: ["bundleName", "bundle_name"])?
                .trimmingCharacters(in: .whitespacesAndNewlines).nonEmpty
                ?? bundleURL.components(separatedBy: "/").last?.nonEmpty
        }

        return LynxPageRequest(
            bundleURL: bundleURL,
            lynxAppId: otaBundleName == nil ? nil : otaAppId,
            bundleName: otaBundleName,
            routeKey: first(query, keys: ["routeKey", "route_key"])?.nonEmpty
                ?? otaBundleName
                ?? bundleURL,
            title: title,
            initialData: initialData,
            globalProps: globalProps,
            fullscreen: fullscreen,
            showNavigationBar: showNavigationBar,
            hideStatusBar: hideStatusBar,
            backGestureEnabled: backGestureEnabled,
            allowHTTPInDebug: allowHTTPInDebug,
            orientation: orientation,
            backgroundColor: backgroundColor,
            widthInPhysicalPixels: widthInPhysicalPixels,
            heightInPhysicalPixels: heightInPhysicalPixels
        )
    }

    private static func applying(
        options: [String: Any],
        to base: LynxPageRequest
    ) throws -> LynxPageRequest {
        let fullscreenOption = firstBool(options, keys: ["fullscreen"])
        let fullscreen = fullscreenOption ?? base.fullscreen

        var showNavigationBar = base.showNavigationBar
        if fullscreenOption != nil {
            showNavigationBar = fullscreen ? false : (base.fullscreen ? true : base.showNavigationBar)
        }
        if let visible = firstBool(options, keys: ["showNavigationBar", "showToolbar"]) {
            showNavigationBar = !fullscreen && visible
        }
        if let hidden = firstBool(
            options,
            keys: ["hidden_nav", "hide_nav_bar", "hideNavigationBar"]
        ) {
            showNavigationBar = !fullscreen && !hidden
        }

        // fullscreen 只控制 edge-to-edge；隐藏状态栏必须使用独立参数显式声明。
        var hideStatusBar = base.hideStatusBar
        if let hidden = firstBool(options, keys: ["hideStatusBar", "hide_status_bar"]) {
            hideStatusBar = hidden
        }
        let initialData = try firstObject(
            options,
            keys: ["initData", "initialData", "initial_data"],
            fallback: base.initialData,
            field: "initData"
        )
        let globalProps = try firstObject(
            options,
            keys: ["globalProps", "global_props"],
            fallback: base.globalProps,
            field: "globalProps"
        )
        let allowHTTPInDebug = firstBool(
            options,
            keys: ["allowHTTPInDebug", "allowHttpInDebug", "allow_http_in_debug"]
        ) ?? base.allowHTTPInDebug
        let orientation = firstString(options, keys: ["orientation", "screen_orientation"])
            .map(parseOrientation) ?? base.orientation
        let backgroundColor = normalizeColor(
            firstString(
                options,
                keys: ["backgroundColor", "background_color", "container_bg_color"]
            )
                ?? base.backgroundColor
        )
        let widthInPhysicalPixels = firstNumber(options, keys: ["width", "width_px"])
            .map { CGFloat(truncating: $0) } ?? base.widthInPhysicalPixels
        let heightInPhysicalPixels = firstNumber(options, keys: ["height", "height_px"])
            .map { CGFloat(truncating: $0) } ?? base.heightInPhysicalPixels
        let requestedAppId = firstString(options, keys: ["lynxAppId", "appId", "lynx_app_id"])
            .flatMap { $0.trimmingCharacters(in: .whitespacesAndNewlines).nonEmpty }
            ?? base.lynxAppId
        let requestedBundleName = firstString(options, keys: ["bundleName", "bundle_name"])
            .flatMap { $0.trimmingCharacters(in: .whitespacesAndNewlines).nonEmpty }
        let otaAppId = RemoteBundlePolicy.isRemote(base.bundleURL) ? nil : requestedAppId
        let otaBundleName = otaAppId.flatMap { _ in
            requestedBundleName
                ?? base.bundleName
                ?? base.bundleURL.components(separatedBy: "/").last?.nonEmpty
        }

        return LynxPageRequest(
            bundleURL: base.bundleURL,
            lynxAppId: otaBundleName == nil ? nil : otaAppId,
            bundleName: otaBundleName,
            routeKey: firstString(options, keys: ["routeKey", "route_key"])?.nonEmpty
                ?? otaBundleName
                ?? base.routeKey,
            title: firstString(options, keys: ["title"])?.nonEmpty ?? base.title,
            initialData: initialData,
            globalProps: globalProps,
            fullscreen: fullscreen,
            showNavigationBar: showNavigationBar,
            hideStatusBar: hideStatusBar,
            backGestureEnabled: firstBool(
                options,
                keys: ["backGestureEnabled", "back_gesture_enabled"]
            ) ?? base.backGestureEnabled,
            allowHTTPInDebug: allowHTTPInDebug,
            orientation: orientation,
            backgroundColor: backgroundColor,
            widthInPhysicalPixels: widthInPhysicalPixels,
            heightInPhysicalPixels: heightInPhysicalPixels
        )
    }

    private static func defaultRequest(bundleURL: String) -> LynxPageRequest {
        LynxPageRequest(
            bundleURL: bundleURL,
            routeKey: bundleURL,
            title: "Lynx",
            initialData: [:],
            globalProps: [:],
            fullscreen: true,
            showNavigationBar: false,
            hideStatusBar: false,
            backGestureEnabled: true,
            allowHTTPInDebug: false,
            orientation: .system,
            backgroundColor: "#FFFFFF",
            widthInPhysicalPixels: nil,
            heightInPhysicalPixels: nil
        )
    }

    private static func decodeObject(_ json: String, field: String) throws -> [String: Any] {
        guard let data = json.data(using: .utf8),
              let value = try? JSONSerialization.jsonObject(with: data),
              let object = value as? [String: Any] else {
            throw LynxRouteError.invalidJSON(field)
        }
        return object
    }

    private static func firstObject(
        _ values: [String: Any],
        keys: [String],
        fallback: [String: Any],
        field: String
    ) throws -> [String: Any] {
        guard let value = firstValue(values, keys: keys) else { return fallback }
        if let object = value as? [String: Any] { return object }
        if let json = value as? String { return try decodeObject(json, field: field) }
        throw LynxRouteError.invalidJSON(field)
    }

    private static func normalizeBundle(_ value: String) -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.contains("://") || trimmed.lowercased().hasPrefix(explorerLocalPrefix) {
            return trimmed
        }
        let path = stripRelativePrefix(trimmed)
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        return path.lowercased().hasPrefix("bundles/")
            ? "assets://Bundles/\(path.split(separator: "/").dropFirst().joined(separator: "/"))"
            : "assets://Bundles/\(path)"
    }

    private static func stripRelativePrefix(_ value: String) -> String {
        var result = value.replacingOccurrences(of: "\\", with: "/")
        while result.hasPrefix("./") { result.removeFirst(2) }
        return result
    }

    private static func normalizeColor(_ value: String) -> String {
        value.hasPrefix("#") ? value : "#\(value)"
    }

    private static func parseOrientation(_ value: String?) -> LynxPageRequest.Orientation {
        switch value?.lowercased() {
        case "portrait", "vertical": return .portrait
        case "landscape", "horizontal": return .landscape
        default: return .system
        }
    }

    private static func parseEncodedQuery(_ value: String) -> [String: String] {
        var result: [String: String] = [:]
        for item in value.split(separator: "&", omittingEmptySubsequences: true) {
            let pieces = item.split(separator: "=", maxSplits: 1, omittingEmptySubsequences: false)
            let rawKey = String(pieces.first ?? "")
            let rawValue = pieces.count > 1 ? String(pieces[1]) : ""
            let key = rawKey.removingPercentEncoding ?? rawKey
            let decoded = rawValue.removingPercentEncoding ?? rawValue
            result[key] = decoded
        }
        return result
    }

    private static func first(_ values: [String: String], keys: [String]) -> String? {
        for key in keys {
            if let value = values[key] { return value }
        }
        return nil
    }

    private static func firstValue(_ values: [String: Any], keys: [String]) -> Any? {
        for key in keys {
            if let value = values[key], !(value is NSNull) { return value }
        }
        return nil
    }

    private static func firstString(_ values: [String: Any], keys: [String]) -> String? {
        guard let value = firstValue(values, keys: keys) else { return nil }
        if let string = value as? String { return string }
        return String(describing: value)
    }

    private static func firstNumber(_ values: [String: Any], keys: [String]) -> NSNumber? {
        guard let value = firstValue(values, keys: keys) else { return nil }
        if let number = value as? NSNumber { return number }
        if let string = value as? String, let double = Double(string) { return NSNumber(value: double) }
        return nil
    }

    private static func firstBool(_ values: [String: Any], keys: [String]) -> Bool? {
        guard let value = firstValue(values, keys: keys) else { return nil }
        if let boolValue = value as? Bool { return boolValue }
        if let number = value as? NSNumber { return number.boolValue }
        if let string = value as? String {
            if ["1", "true", "yes"].contains(string.lowercased()) { return true }
            if ["0", "false", "no"].contains(string.lowercased()) { return false }
        }
        return nil
    }

    private static func bool(_ value: String?, default fallback: Bool) -> Bool {
        guard let value else { return fallback }
        return ["1", "true", "yes"].contains(value.lowercased())
    }
}

private extension String {
    var nonEmpty: String? { isEmpty ? nil : self }
}

File: android/lynx-shell/src/main/java/com/example/lynxshell/routing/LynxRouteParser.kt
package com.example.lynxshell.routing

import android.content.Intent
import android.net.Uri
import com.example.lynxshell.BuildConfig
import com.example.lynxshell.model.KeyboardBehavior
import com.example.lynxshell.model.LynxPageRequest
import com.example.lynxshell.model.PageOrientation
import org.json.JSONObject

/**
 * 把 Intent、深链、Explorer 本地地址和 Sparkling hybrid URL 统一转换成 LynxPageRequest。
 *
 * 兼容的典型参数别名：
 * - `hidden_nav` / `hide_nav_bar` / `hideNavigationBar`
 * - `orientation` / `screen_orientation`
 * - `initData` / `initialData` / `initial_data`
 * - `globalProps` / `global_props`
 */
object LynxRouteParser {
    private const val EXPLORER_LOCAL_PREFIX = "file://lynx?local://"

    fun parse(intent: Intent): Result<LynxPageRequest> = runCatching {
        intent.getStringExtra(LynxPageRequest.EXTRA_BUNDLE_URL)?.let {
            return@runCatching fromIntentExtras(intent).validated()
        }

        val data = requireNotNull(intent.data) { "Intent 缺少 Lynx 页面参数" }
        fromUri(data).validated()
    }

    /** Native Module 打开页面时使用；optionsJSON 必须是 JSON Object。 */
    fun fromBridge(url: String, optionsJson: String): LynxPageRequest {
        val options = if (optionsJson.isBlank()) JSONObject() else JSONObject(optionsJson)
        val base = if (url.contains("://")) {
            fromRawRoute(url)
        } else {
            defaultRequest(normalizeBundle(url))
        }
        return applyOptions(base, options).validated()
    }

    fun fromUri(uri: Uri): LynxPageRequest = fromRawRoute(uri.toString())

    private fun fromRawRoute(rawRoute: String): LynxPageRequest {
        if (rawRoute.startsWith(EXPLORER_LOCAL_PREFIX, ignoreCase = true)) {
            return requestFromExplorerLocal(rawRoute)
        }

        val uri = Uri.parse(rawRoute)
        val scheme = uri.scheme.orEmpty().lowercase()
        val query = parseEncodedQuery(uri.encodedQuery)

        if (scheme == "lynxshell" || (scheme == "hybrid" && uri.host == "lynxview_page")) {
            val rawBundle = first(query, "url", "bundle")
                ?.takeIf { it.isNotBlank() }
                ?: throw IllegalArgumentException("路由缺少 url 或 bundle 参数")
            return requestFromQuery(normalizeBundle(rawBundle), query)
        }

        if (scheme in setOf("assets", "https", "http")) {
            return requestFromQuery(rawRoute, query)
        }

        throw IllegalArgumentException("不支持的路由: $rawRoute")
    }

    /**
     * Explorer 的 `file://lynx?local://xxx.bundle?...` 不是标准 URL：第一个问号属于协议，
     * 第二个问号才开始承载页面参数，因此不能直接依赖 Uri.getQueryParameter。
     */
    private fun requestFromExplorerLocal(rawRoute: String): LynxPageRequest {
        val payload = rawRoute.substring(EXPLORER_LOCAL_PREFIX.length)
        val optionIndex = payload.indexOf('?')
        val encodedPath = if (optionIndex >= 0) payload.substring(0, optionIndex) else payload
        val encodedOptions = if (optionIndex >= 0) payload.substring(optionIndex + 1) else ""
        val path = Uri.decode(encodedPath).trim()
        require(path.isNotBlank()) { "Explorer 本地 Bundle 路径为空" }
        val bundle = EXPLORER_LOCAL_PREFIX + stripRelativePrefix(path)
        return requestFromQuery(bundle, parseEncodedQuery(encodedOptions))
    }

    private fun fromIntentExtras(intent: Intent): LynxPageRequest {
        // Bundle 默认 edge-to-edge 且不显示宿主 Toolbar；状态栏保留并透明覆盖在内容上。
        val fullscreen = intent.getBooleanExtra(LynxPageRequest.EXTRA_FULLSCREEN, true)
        val showToolbar = if (intent.hasExtra(LynxPageRequest.EXTRA_SHOW_TOOLBAR)) {
            intent.getBooleanExtra(LynxPageRequest.EXTRA_SHOW_TOOLBAR, !fullscreen)
        } else {
            !fullscreen
        }
        val hideStatusBar = if (intent.hasExtra(LynxPageRequest.EXTRA_HIDE_STATUS_BAR)) {
            intent.getBooleanExtra(LynxPageRequest.EXTRA_HIDE_STATUS_BAR, false)
        } else {
            false
        }

        return LynxPageRequest(
            bundleUrl = intent.getStringExtra(LynxPageRequest.EXTRA_BUNDLE_URL).orEmpty(),
            lynxAppId = intent.getStringExtra(LynxPageRequest.EXTRA_LYNX_APP_ID),
            bundleName = intent.getStringExtra(LynxPageRequest.EXTRA_BUNDLE_NAME),
            routeKey = intent.getStringExtra(LynxPageRequest.EXTRA_ROUTE_KEY).orEmpty(),
            title = intent.getStringExtra(LynxPageRequest.EXTRA_TITLE) ?: "Lynx",
            initDataJson = intent.getStringExtra(LynxPageRequest.EXTRA_INIT_DATA) ?: "{}",
            globalPropsJson = intent.getStringExtra(LynxPageRequest.EXTRA_GLOBAL_PROPS) ?: "{}",
            fullscreen = fullscreen,
            showToolbar = showToolbar,
            hideStatusBar = hideStatusBar,
            backGestureEnabled = intent.getBooleanExtra(
                LynxPageRequest.EXTRA_BACK_GESTURE_ENABLED,
                true,
            ),
            allowHttpInDebug = intent.getBooleanExtra(LynxPageRequest.EXTRA_ALLOW_HTTP, false),
            orientation = enumValueOrDefault(
                intent.getStringExtra(LynxPageRequest.EXTRA_ORIENTATION),
                PageOrientation.SYSTEM,
            ),
            keyboardBehavior = parseKeyboardBehavior(
                intent.getStringExtra(LynxPageRequest.EXTRA_KEYBOARD_BEHAVIOR),
            ),
            backgroundColor = intent.getStringExtra(LynxPageRequest.EXTRA_BACKGROUND_COLOR) ?: "#FFFFFF",
            widthPx = intent.optionalInt(LynxPageRequest.EXTRA_WIDTH),
            heightPx = intent.optionalInt(LynxPageRequest.EXTRA_HEIGHT),
            density = intent.optionalFloat(LynxPageRequest.EXTRA_DENSITY),
        )
    }

    private fun requestFromQuery(bundleUrl: String, query: Map<String, String>): LynxPageRequest {
        // 深链未声明 chrome 参数时同样采用纯 Lynx 全屏容器。
        val fullscreen = bool(first(query, "fullscreen"), true)
        val hiddenNavigation = bool(
            first(query, "hidden_nav", "hide_nav_bar", "hideNavigationBar"),
            false,
        )
        val explicitNavigation = first(query, "showNavigationBar", "showToolbar")
        val showToolbar = if (explicitNavigation != null) {
            !fullscreen && bool(explicitNavigation, true)
        } else {
            !fullscreen && !hiddenNavigation
        }
        val hideStatusBar = bool(
            first(query, "hide_status_bar", "hideStatusBar"),
            false,
        )

        // HTTPS 是独立的 Direct Remote 语义；即使调用方误带 OTA 字段，也不能进入本地
        // current/Manifest 链路。只有受控的 assets/local Bundle 才允许携带 appId。
        val isDirectRemote = LynxPageRequest.isRemoteBundleUrl(bundleUrl)
        val otaAppId = first(query, "lynxAppId", "appId", "lynx_app_id")
            ?.takeUnless { isDirectRemote }
            ?.trim()
            ?.takeIf { it.isNotBlank() }
        // Scheme 中传 appId + bundleName 时进入 OTA；没有 appId 仍保持原来的直加载语义。
        val otaBundleName = if (otaAppId != null) {
            first(query, "bundleName", "bundle_name")
                ?.trim()
                ?.takeIf { it.isNotBlank() }
                ?: bundleUrl
                    .takeUnless { LynxPageRequest.isRemoteBundleUrl(it) }
                    ?.removePrefix("assets://")
                    ?.removePrefix("bundles/")
        } else {
            null
        }

        return LynxPageRequest(
            bundleUrl = bundleUrl,
            lynxAppId = otaAppId?.takeIf { otaBundleName != null },
            bundleName = otaBundleName,
            routeKey = first(query, "routeKey", "route_key") ?: otaBundleName ?: bundleUrl,
            title = first(query, "title")?.takeIf { it.isNotBlank() } ?: "Lynx",
            initDataJson = first(query, "initData", "initialData", "initial_data") ?: "{}",
            // Playground 的导航、主题和容器示例都从 __globalProps.queryItems 读取路由参数。
            globalPropsJson = mergeQueryIntoGlobalProps(
                first(query, "globalProps", "global_props") ?: "{}",
                query,
            ),
            fullscreen = fullscreen,
            showToolbar = showToolbar,
            hideStatusBar = hideStatusBar,
            backGestureEnabled = bool(
                first(query, "backGestureEnabled", "back_gesture_enabled"),
                true,
            ),
            allowHttpInDebug = BuildConfig.DEBUG && bool(
                first(query, "allowHttpInDebug", "allow_http_in_debug"),
                false,
            ),
            orientation = parseOrientation(first(query, "orientation", "screen_orientation")),
            keyboardBehavior = parseKeyboardBehavior(
                first(query, "keyboardBehavior", "keyboard_behavior"),
            ),
            backgroundColor = normalizeColor(
                first(query, "backgroundColor", "background_color", "container_bg_color")
                    ?: "#FFFFFF",
            ),
            widthPx = first(query, "width", "width_px")?.toIntOrNull(),
            heightPx = first(query, "height", "height_px")?.toIntOrNull(),
            density = first(query, "density")?.toFloatOrNull()?.let { value ->
                // Explorer URL 中 density 通常表示 dpi；小于 10 时按 Android density 直接处理。
                if (value >= 10f) value / 160f else value
            },
        )
    }

    private fun applyOptions(base: LynxPageRequest, options: JSONObject): LynxPageRequest {
        val fullscreenOption = options.firstBoolean("fullscreen")
        val fullscreen = fullscreenOption ?: base.fullscreen

        var showToolbar = base.showToolbar
        if (fullscreenOption != null) {
            // 从全屏切回普通页面时恢复合理的原生导航默认值。
            showToolbar = if (fullscreen) false else if (base.fullscreen) true else base.showToolbar
        }
        options.firstBoolean("showNavigationBar", "showToolbar")?.let { visible ->
            showToolbar = !fullscreen && visible
        }
        options.firstBoolean("hidden_nav", "hide_nav_bar", "hideNavigationBar")?.let { hidden ->
            showToolbar = !fullscreen && !hidden
        }

        // fullscreen 只控制 edge-to-edge；是否隐藏状态栏必须由独立参数明确表达。
        var hideStatusBar = base.hideStatusBar
        options.firstBoolean("hideStatusBar", "hide_status_bar")?.let { hideStatusBar = it }

        val keyboardBehavior = options.firstString("keyboardBehavior", "keyboard_behavior")
            ?.let(::parseKeyboardBehavior)
            ?: base.keyboardBehavior

        val isDirectRemote = LynxPageRequest.isRemoteBundleUrl(base.bundleUrl)
        val appId = options.firstString("lynxAppId", "appId", "lynx_app_id")
            ?.trim()
            ?.takeIf { it.isNotBlank() }
            ?.takeUnless { isDirectRemote }
            ?: base.lynxAppId?.takeUnless { isDirectRemote }
        val requestedBundleName = options.firstString("bundleName", "bundle_name")
            ?.trim()
            ?.takeIf { it.isNotBlank() }
        val bundleName = if (isDirectRemote) {
            null
        } else if (requestedBundleName != null) {
            requestedBundleName
        } else if (base.bundleName != null) {
            base.bundleName
        } else if (
            appId != null && !LynxPageRequest.isRemoteBundleUrl(base.bundleUrl)
        ) {
            base.bundleUrl.substringAfterLast('/').substringBefore('?')
        } else {
            null
        }

        return base.copy(
            lynxAppId = appId?.takeIf { bundleName != null },
            bundleName = bundleName,
            routeKey = options.firstString("routeKey", "route_key")
                ?.takeIf { it.isNotBlank() }
                ?: base.routeKey,
            title = options.firstString("title")?.takeIf { it.isNotBlank() } ?: base.title,
            initDataJson = options.firstObjectJson(
                keys = arrayOf("initData", "initialData", "initial_data"),
                fallback = base.initDataJson,
            ),
            globalPropsJson = options.firstObjectJson(
                keys = arrayOf("globalProps", "global_props"),
                fallback = base.globalPropsJson,
            ),
            fullscreen = fullscreen,
            showToolbar = showToolbar,
            hideStatusBar = hideStatusBar,
            backGestureEnabled = options.firstBoolean(
                "backGestureEnabled",
                "back_gesture_enabled",
            ) ?: base.backGestureEnabled,
            allowHttpInDebug = BuildConfig.DEBUG && (
                options.firstBoolean("allowHttpInDebug", "allow_http_in_debug")
                    ?: base.allowHttpInDebug
                ),
            orientation = options.firstString("orientation", "screen_orientation")
                ?.let(::parseOrientation)
                ?: base.orientation,
            keyboardBehavior = keyboardBehavior,
            backgroundColor = normalizeColor(
                options.firstString(
                    "backgroundColor",
                    "background_color",
                    "container_bg_color",
                ) ?: base.backgroundColor,
            ),
            widthPx = options.firstNumber("width", "width_px")?.toInt() ?: base.widthPx,
            heightPx = options.firstNumber("height", "height_px")?.toInt() ?: base.heightPx,
            density = options.firstNumber("density")?.toFloat()?.let { value ->
                if (value >= 10f) value / 160f else value
            } ?: base.density,
        )
    }

    private fun defaultRequest(bundleUrl: String) = LynxPageRequest(bundleUrl = bundleUrl)

    private fun normalizeBundle(value: String): String {
        val trimmed = value.trim()
        if (trimmed.contains("://") || trimmed.startsWith(EXPLORER_LOCAL_PREFIX)) return trimmed
        val path = stripRelativePrefix(trimmed).trimStart('/')
        return if (path.startsWith("bundles/", ignoreCase = true)) {
            "assets://$path"
        } else {
            "assets://bundles/$path"
        }
    }

    private fun stripRelativePrefix(value: String): String {
        var result = value.replace('\\', '/')
        while (result.startsWith("./")) result = result.removePrefix("./")
        return result
    }

    private fun parseOrientation(value: String?): PageOrientation = when (value?.lowercase()) {
        "portrait", "vertical" -> PageOrientation.PORTRAIT
        "landscape", "horizontal" -> PageOrientation.LANDSCAPE
        else -> PageOrientation.SYSTEM
    }

    private fun parseKeyboardBehavior(value: String?): KeyboardBehavior = when (value?.trim()?.lowercase()) {
        "resize" -> KeyboardBehavior.RESIZE
        "pan" -> KeyboardBehavior.PAN
        "nothing", "none" -> KeyboardBehavior.NOTHING
        else -> KeyboardBehavior.SYSTEM
    }

    private fun normalizeColor(value: String): String =
        if (value.startsWith("#")) value else "#$value"

    private fun parseEncodedQuery(encodedQuery: String?): Map<String, String> {
        if (encodedQuery.isNullOrBlank()) return emptyMap()
        val result = linkedMapOf<String, String>()
        encodedQuery.split('&').forEach { part ->
            if (part.isBlank()) return@forEach
            val pieces = part.split('=', limit = 2)
            val key = Uri.decode(pieces[0])
            val value = Uri.decode(pieces.getOrElse(1) { "" })
            result[key] = value
        }
        return result
    }

    private fun first(query: Map<String, String>, vararg keys: String): String? =
        keys.firstNotNullOfOrNull(query::get)

    private fun bool(value: String?, fallback: Boolean): Boolean {
        if (value == null) return fallback
        return value == "1" || value.equals("true", ignoreCase = true) ||
            value.equals("yes", ignoreCase = true)
    }

    private fun mergeQueryIntoGlobalProps(
        globalPropsJson: String,
        query: Map<String, String>,
    ): String = JSONObject(globalPropsJson).apply {
        put("queryItems", JSONObject(query))
    }.toString()

    private fun JSONObject.firstValue(vararg keys: String): Any? {
        for (key in keys) {
            if (has(key) && !isNull(key)) return get(key)
        }
        return null
    }

    private fun JSONObject.firstString(vararg keys: String): String? =
        firstValue(*keys)?.let { value -> if (value is String) value else value.toString() }

    private fun JSONObject.firstNumber(vararg keys: String): Number? = when (val value = firstValue(*keys)) {
        is Number -> value
        is String -> value.toDoubleOrNull()
        else -> null
    }

    private fun JSONObject.firstBoolean(vararg keys: String): Boolean? = when (val value = firstValue(*keys)) {
        is Boolean -> value
        is Number -> value.toInt() != 0
        is String -> when {
            value == "1" || value.equals("true", true) || value.equals("yes", true) -> true
            value == "0" || value.equals("false", true) || value.equals("no", true) -> false
            else -> null
        }
        else -> null
    }

    private fun JSONObject.firstObjectJson(keys: Array<String>, fallback: String): String {
        return when (val value = firstValue(*keys)) {
            null -> fallback
            is JSONObject -> value.toString()
            is String -> value
            else -> throw IllegalArgumentException("${keys.first()} 必须是 JSON Object 或其字符串形式")
        }
    }

    private fun Intent.optionalInt(key: String): Int? = if (hasExtra(key)) getIntExtra(key, 0) else null
    private fun Intent.optionalFloat(key: String): Float? = if (hasExtra(key)) getFloatExtra(key, 0f) else null

    private inline fun <reified T : Enum<T>> enumValueOrDefault(value: String?, fallback: T): T =
        runCatching { enumValueOf<T>(value.orEmpty()) }.getOrDefault(fallback)
}
